### DISCUSSION TITLE
delete elements with jsonpath

### DISCUSSION THREAD

**Discussion by romange** [danielaparker/jsoncons#379] [Q&A] [answered]
Consider the following json:
`auto j1 = R"({"c":{"a":1, "b":2}, "d":{"a":1, "b":2, "c":3}, "e": [1,2]})"_json`;

I would like to implement something like this
```cpp
auto elems = jsonpath::json_query(j1, "$.d.*", result_options::path);
auto j2 = Delete(j1, elems);
/* j2 should be {"c":{"a":1, "b":2}, "d":{}, "e": [1,2]} */
```

what's missing is `Delete` functionality. Is there an easy way to implement it, maybe using other jsoncons APIs?

**Comment by danielaparker** [comment, marked as answer]
You could do it like this:

```cpp
#include <jsoncons/json.hpp>
#include <jsoncons_ext/jsonpath/jsonpath.hpp>

using namespace jsoncons;

int main()
{
    auto j = json::parse(R"({"c":{"a":1, "b":2}, "d":{"a":1, "b":2, "c":3}, "e": [1,2]})");

    auto deleter = [](const json::string_view_type& path, json& val) 
                      {if (val.is_object()) val.erase(val.object_range().begin(), val.object_range().end());};

    jsonpath::json_replace(j, "$.d", deleter);

    std::cout << j << std::endl;
}
```

Output:

```json
{"c":{"a":1,"b":2},"d":{},"e":[1,2]}
```

**Comment by hegsie** [comment]
Hi,

I'm trying to re-use this example and I am getting the below error, is this recent feature (do I need to update my header files)?

1>C:\src\WixJsonExtension\src\ca\jsoncons\json_type_traits.hpp(75,45): error C2338: static_assert failed: 'to_json not implemented'
1>C:\src\WixJsonExtension\src\ca\jsoncons\json_type_traits.hpp(74,1): message : while compiling class template member function 'Json jsoncons::json_type_traits<Json,T,void>::to_json(const T &,const std::allocator<char> &)'
1>        with
1>        [
1>            Json=jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>,
1>            T=DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>
1>        ]
1>C:\src\WixJsonExtension\src\ca\jsoncons\basic_json.hpp(3291,15): message : see reference to function template instantiation 'Json jsoncons::json_type_traits<Json,T,void>::to_json(const T &,const std::allocator<char> &)' being compiled
1>        with
1>        [
1>            Json=jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>,
1>            T=DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>
1>        ]
1>C:\src\WixJsonExtension\src\ca\jsoncons\basic_json.hpp(3291,45): message : see reference to class template instantiation 'jsoncons::json_type_traits<jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>,T,void>' being compiled
1>        with
1>        [
1>            T=DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>
1>        ]
1>C:\src\WixJsonExtension\src\ca\jsoncons_ext\jsonpath\json_query.hpp(610,1): message : see reference to function template instantiation 'jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>> &jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>::operator =<DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>>(const T &)' being compiled
1>        with
1>        [
1>            T=DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>
1>        ]
1>C:\src\WixJsonExtension\src\ca\jsoncons_ext\jsonpath\json_query.hpp(610,49): message : see reference to function template instantiation 'jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>> &jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>::operator =<DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>>(const T &)' being compiled
1>        with
1>        [
1>            T=DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>
1>        ]
1>C:\src\WixJsonExtension\src\ca\jsoncons_ext\jsonpath\json_query.hpp(139,1): message : see reference to function template instantiation 'void jsoncons::jsonpath::detail::jsonpath_evaluator<Json,Json &,jsoncons::jsonpath::detail::VoidPathConstructor<Json>>::replace<DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>&>(T)' being compiled
1>        with
1>        [
1>            Json=jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>,
1>            T=DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b> &
1>        ]
1>C:\src\WixJsonExtension\src\ca\jsoncons_ext\jsonpath\json_query.hpp(139,26): message : see reference to function template instantiation 'void jsoncons::jsonpath::detail::jsonpath_evaluator<Json,Json &,jsoncons::jsonpath::detail::VoidPathConstructor<Json>>::replace<DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>&>(T)' being compiled
1>        with
1>        [
1>            Json=jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>,
1>            T=DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b> &
1>        ]
1>C:\src\WixJsonExtension\src\ca\JsonFile.cpp(288,43): message : see reference to function template instantiation 'void jsoncons::jsonpath::json_replace<jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>,DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b>&>(Json &,const std::basic_string_view<char,std::char_traits<char>> &,T)' being compiled
1>        with
1>        [
1>            Json=jsoncons::basic_json<char,jsoncons::sorted_policy,std::allocator<char>>,
1>            T=DeleteJsonPath::<lambda_5d70fdfa3e27f057ebe9fca70143530b> &
1>        ]

**Reply by danielaparker** [reply]
It works for me. What version of jsoncons are you on?

**Comment by hegsie** [comment]
yup looks like a version issue, updated to latest and it works now, thanks
